### PR TITLE
Adding tests for encrypt and minor tweaks for error message + edge cases

### PR DESCRIPTION
**Pull request by gauntface**
@wibblymat @beverloo

This PR adds tests for encrypt method - should run on travis too and adds / changes error logging where appropriate.

### CODE REVIEW

File: src/encrypt.js [hunk 1]
```diff
@@ -41,16 +41,30 @@ function encrypt(message, subscription) {
...
       !subscription.keys.auth) {
-    throw new Error('Subscription has no encryption details');
+    throw new Error('Subscription is missing some encryption details.');
   }
+
   const clientPublicKey = new Buffer(subscription.keys.p256dh, 'base64');
   const clientAuthToken = new Buffer(subscription.keys.auth, 'base64');
 
+  if (clientAuthToken.length !== 16) {
+    throw new Error('Subscriptions Auth token is not 16 bytes.');
```
Comment (wibblymat) [inline src/encrypt.js]: Here and the next if statement: subscription's rather than subscriptions

File: test/index.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      const response = library.encrypt('Hello, World', VALID_SUBSCRIPTION);
+      Buffer.isBuffer(response.ciphertext).should.equal(true);
+      Buffer.isBuffer(response.salt).should.equal(true);
+      response.salt.should.have.length(SALT_LENGTH);
+      Buffer.isBuffer(response.serverPublicKey).should.equal(true);
+      response.serverPublicKey.should.have.length(SERVER_PUBLIC_KEY_LENGTH);
+    });
+
+    it('should throw an error due to subscription with no keys being passed in', function() {
+      const library = require('../src/index.js');
```
Comment (wibblymat) [inline test/index.js:91]: Is there a particular reason to require the library in the tests rather than once at the top?

File: test/index.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    });
+
+    it('should throw an error due to an invalid client public key', function() {
+      const library = require('../src/index.js');
+      expect(
+        () => library.encrypt('Hello, World', INVALID_P256DH_SUBSCRIPTION)
+      ).to.throw('Subscriptions client key (p256dh) is invalid.');
+    });
+
+    it('should not throw an error when no p256dh key is passed in', function() {
```
Comment (wibblymat) [inline test/index.js]: `should not` => `should`

File: test/index.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+          auth: VALID_SUBSCRIPTION.keys.auth
+        }
+      };
+
+      expect(
+        () => library.encrypt('Hello, World', subscription)
+      ).to.throw('Subscription is missing some encryption details');
+    });
+
+    it('should not throw an error when no p256dh key is passed in', function() {
```
Comment (wibblymat) [inline test/index.js]: Looks like an exact copy of the previous test

File: test/index.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+      response.serverPublicKey.should.have.length(SERVER_PUBLIC_KEY_LENGTH);
+      response.serverPublicKey.toString('base64').should.equal(EXAMPLE_SERVER_KEYS.public);
+
+      response.ciphertext.toString('base64').should.equal(EXAMPLE_OUTPUT);
+    });
+
+    it('should throw an error when the input is too large', function() {
+      const library = require('../src/index.js');
+
+      let largeInput = '';
```
Comment (wibblymat) [inline test/index.js]:
```
const largeInput = new Buffer(4081);
largeInput.fill(0);

expect(
  () => library.encrypt(largeInput.toString('utf8'), VALID_SUBSCRIPTION)
).to.throw('Payload is too large. The max number of bytes is 4080, input is 4081 bytes.');
```

Or, if you are confident that the character 'A' in UTF-8 is a single byte:

```
const largeInput = 'A'.repeat(4081);

expect(
  () => library.encrypt(largeInput, VALID_SUBSCRIPTION)
).to.throw('Payload is too large. The max number of bytes is 4080, input is 4081 bytes.');
```

File: test/index.js (hunk 2 above)
Comment (gauntface) [inline test/index.js:91]: yeah in one or two of them I use proxyquire to inject dependency behavior and calling in each one prevents any confusion.

File: test/index.js (hunk 5 above)
Comment (gauntface) [inline test/index.js]: Do you have a preference for A?

I just cared for the reliable buffer length to be honest. Content was irrelevant.

Comment (wibblymat) [inline test/index.js]: No, not really. I was just trying to create a reliably 4081 byte buffer in less code, and picked what you picked :)

At the very least the loop can be replaced with `let largeInput = 'A'.repeat(4081);`

Comment (gauntface) [inline test/index.js]: Sorry, I totally didn't see what you were getting at. Just used fill(0) with buffer size, much nicer.

**Comment by gauntface** [comment]
@wibblymat happy?

File: test/index.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    });
+
+    it('should throw an error when the input is too large', function() {
+      const library = require('../src/index.js');
+
+      let largeInput = new Buffer(4081);
+      largeInput.fill(0);
+
+      expect(
+        () => library.encrypt(new Buffer(largeInput).toString('utf8'), VALID_SUBSCRIPTION)
```
Comment (wibblymat) [inline test/index.js]: One last thing - `largeInput` is now a Buffer already, so no need to call `new Buffer()` here and in next `expect`

**Comment by wibblymat** [comment]
Sweeet, thanks!
